### PR TITLE
Add basic ping testsuite to be used to verify MM setup

### PR DESCRIPTION
**Pull request by asdil12**
https://progress.opensuse.org/issues/138302
VR: https://openqa.opensuse.org/tests/3864203
Schedule: https://github.com/os-autoinst/opensuse-jobgroups/pull/413

### CODE REVIEW

**Review by Martchus — APPROVED** [review]

File: tests/network/ping.pm [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    if (get_var('IS_MM_CLIENT')) {
+        my $remote = '10.0.2.101';
+        my $size = get_var('MM_MTU', 1380) - 28;    # 8 bytes ICMP and 20 bytes IPv4 header
+        script_retry("ping -M do -s $size -c1 -W1 $remote", retry => 5, delay => 2, fail_message => "Failed to ping $remote");
+    }
+
+    # Finish job
+    wait_for_children if (get_var('IS_MM_SERVER'));
+
```
Comment (okurz) [inline tests/network/ping.pm:38]:
```suggestion
use constant WHATEVER_1380_MEANS => 1380;
use constant ICMP_IPV4_HEADER_SIZE => 28;  # 8 bytes ICMP and 20 bytes IPv4 header
…
    mutex_wait 'barrier_setup_done';
    select_serial_terminal;
    barrier_wait 'SETUP_DONE';
    if (get_var('IS_MM_CLIENT')) {
        my $remote = '10.0.2.101';
        my $size = get_var('MM_MTU', WHATEVER_1380_MEANS) - ICMP_IPV4_HEADER_SIZE;
        script_retry("ping -M do -s $size -c1 -W1 $remote", retry => 5, delay => 2, fail_message => "Failed to ping '$remote'");
    }
    # Finish job
    wait_for_children if get_var('IS_MM_SERVER');
```

and why the `script_retry`?

File: schedule/functional/ping.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+name:           ping
+description:    >
+    Maintainer: dheidler.
```
Comment (okurz) [inline schedule/functional/ping.yaml:3]:
```suggestion
    Maintainer: dheidler
```

File: tests/network/ping.pm [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# SUSE's openQA tests
+#
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Package: iputils
+# Summary: Ping another machine
```
Comment (okurz) [inline tests/network/ping.pm:7]: I suggest to mention the purpose for the custom size

**Review by okurz — CHANGES_REQUESTED** [review]

File: tests/network/ping.pm (hunk 1 above)
Comment (foursixnine) [reply tests/network/ping.pm:38]: > and why the script_retry?

Actually, why not? We want to know life is good, so not dying on the first ping is a good idea, perhaps 5 is a bit too much, but isn't the end of the world. 

> WHATEVER_1380_MEANS

I guess that's the MTU packet size

File: tests/network/ping.pm [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    mutex_wait 'barrier_setup_done';
+
+    select_serial_terminal;
+    barrier_wait 'SETUP_DONE';
+
+    if (get_var('IS_MM_CLIENT')) {
+        my $remote = '10.0.2.101';
+        my $size = get_var('MM_MTU', 1380) - 28;    # 8 bytes ICMP and 20 bytes IPv4 header
+        script_retry("ping -M do -s $size -c1 -W1 $remote", retry => 5, delay => 2, fail_message => "Failed to ping $remote");
```
Comment (foursixnine) [inline tests/network/ping.pm:33]: This is duplicating `utils::ping_size_check`

File: tests/network/ping.pm (hunk 3 above)
Comment (foursixnine) [inline tests/network/ping.pm:7]: This leads one to believe that this is a new test for ping, and we already have one of those, one can argue that the tests have different purposes, and in the end I see no point in adding one more MM scenario, just for the sake of it.

**Review by foursixnine — CHANGES_REQUESTED** [review]
Please extract `network::ping::run` and move it to a `verify_setup` or whatever you prefer to call it on `setup_multimachine`

File: tests/network/ping.pm (hunk 4 above)
Comment (asdil12) [reply tests/network/ping.pm:33]: No it is not.
`ping_size_check` does `assert_script_run´ while I do `script_retry` to make this test more robust.

File: tests/network/ping.pm (hunk 3 above)
Comment (asdil12) [reply tests/network/ping.pm:7]: This test is required to be cloned in a nested test to check our MM worker setup scripts in the openQA test distribution.
It also makes sense to run it on tw to ensure that it works outside of the nested openQA-in-openQA environment.

File: tests/network/ping.pm (hunk 1 above)
Comment (asdil12) [reply tests/network/ping.pm:38]: I think the comment  should be descriptive enough.

File: tests/network/ping.pm (hunk 3 above)
Comment (asdil12) [reply tests/network/ping.pm]: > we already have one of those

At least I didn't find a MM ping test.

Comment (asdil12) [reply tests/network/ping.pm]: Updated Summary.

File: schedule/functional/ping.yaml (hunk 2 above)
Comment (asdil12) [reply schedule/functional/ping.yaml]: Updated description.

**Review by b10n1k — CHANGES_REQUESTED** [review]
`IS_MM_CLIENT` and `IS_MM_SERVER` needs to be added in the https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/variables.md

Also consider to the mutex. they dont seem to add anything. Barriers should be enough

**Comment by asdil12** [comment]
> Also consider to the mutex. they dont seem to add anything. Barriers should be enough

I used the same code as the wireguard test.
I guess the wireguard test used a mutex to secure the barrier creation as barriers might not block, when waiting on an uncreated barrier but mutexes do. 

At least there is no behaviour documented for waiting on barriers before they got created.

File: tests/network/ping.pm (hunk 4 above)
Comment (foursixnine) [reply tests/network/ping.pm:33]: then backport the changes, taking into account the behavior that ping_size_check does.

**Comment by asdil12** [comment]
> Please extract network::ping::run and move it to a verify_setup or whatever you prefer to call it on setup_multimachine

@foursixnine I would still need this testsuite - even if it doesn't have it's own ping module.

**Comment by asdil12** [comment]
@b10n1k had another look and all MM tests seem to use a mutex to guard the barrier creation.

**Comment by asdil12** [comment]
New VR: https://openqa.opensuse.org/tests/3865456

File: schedule/functional/ping.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+name:           ping
```
Comment (okurz) [inline schedule/functional/ping.yaml:1]: How about "multi-machine ping"?

File: tests/network/ping.pm (hunk 3 above)
Comment (asdil12) [reply tests/network/ping.pm]: That this is a ping test? I thought I did.
Anyway this file is not part of the PR anymore as I followed @foursixnine 's suggestion to move it into the setup_multimachine module.

File: tests/network/ping.pm (hunk 4 above)
Comment (asdil12) [reply tests/network/ping.pm]: done.

**Review by okurz — CHANGES_REQUESTED** [review]
see newly added comment about the name. Rest looks fine to me

File: schedule/functional/ping.yaml (hunk 5 above)
Comment (asdil12) [reply schedule/functional/ping.yaml]: renamed to `mm_ping`

Comment (kalikiana) [reply schedule/functional/ping.yaml]: Yeah. No strong opinion on the naming, but I assume it's not `[...]MM[...]as well` as the title suggests. So I'd say drop the `as well` from the title, and call it `multi-machine-ping` or `mm-ping`.

Comment (kalikiana) [reply schedule/functional/ping.yaml]: @asdil12 I guess you missed my comment. The title is misleading

Comment (asdil12) [reply schedule/functional/ping.yaml]: @kalikiana I have no idea what you mean - the string "well" doesn't appear in that file.
```
$ grep well schedule/functional/mm_ping.yaml ; echo $?
1
```

**Review by okurz — APPROVED** [review]

**Review by foursixnine — APPROVED** [review]

**Comment by foursixnine** [comment]
> > Please extract network::ping::run and move it to a verify_setup or whatever you prefer to call it on setup_multimachine
> 
> @foursixnine I would still need this testsuite - even if it doesn't have it's own ping module.

Got it :) all good now

**Review by kalikiana — APPROVED** [review]

**Review by b10n1k — APPROVED** [review]

**Comment by okurz** [comment]
That should suffice

**Pull request merged**